### PR TITLE
PP-12374 Clarify logs with charge status for Wallet payment requests

### PR DESCRIPTION
**Pull request by marcotranchino**
With this change, we are updating the Wallet Authorise Service in order to log
the "charge status" alongside and more prominently than the "request status"
for Apple and Google payments.

This is needed, because the current log message implies that the "success"
status logged is related to the payment itself, e.g. with "APPLE_PAY
authorisation success".

However that "success" word is for the request_status, showing that Connector
managed to connect to the Apple Pay Gateway successfully.

To make the log clearer, avoiding confusion, this commit changes the text
logged, by indicating the charge status first, and then also showing the
request status but in brackets and clearly indicated as such.

This would mean that for the example above, the log will now show:
"APPLE_PAY authorisation - charge status=AUTHORISATION REJECTED, request status=success"

Note that we have verified that the log message being changed is **not used** by the charts displayed in the
[Wallet Payments dashboard](https://gds.splunkcloud.com/en-GB/app/gds-004-pay/googlepay_payments) in Splunk.

Note that by adding the charge status to the parameters passed to the log,
instead of replacing it, we are keeping the metricRegistry and the
walletPaymentAuthorisationSuccessCounter unchanged.

**Further changes will be needed to correct them: we have agreed to coordinate
them around a follow up PR.**

Further information in Jira.

https://payments-platform.atlassian.net/browse/PP-12374

## Examples of failed Apple Pay events logged as success

https://gds.splunkcloud.com/en-GB/app/gds-004-pay/report?s=%2FservicesNS%2Fnobody%2Fgds-004-pay%2Fsaved%2Fsearches%2FFailed%2520Apple%2520Pay%2520payments%2520logged%2520as%2520success&sid=_bWFyY28udHJhbmNoaW5vQGRpZ2l0YWwuY2FiaW5ldC1vZmZpY2UuZ292LnVr_bWFyY28udHJhbmNoaW5vQGRpZ2l0YWwuY2FiaW5ldC1vZmZpY2UuZ292LnVr_Z2RzLTAwNC1wYXk__RMD50c086d415149575d_at_1714139132_85155&display.page.search.mode=smart&dispatch.sample_ratio=1&earliest=-24h%40h&latest=now

## Feedback request

**QUESTION 1**
**Are we happy with the formatting changes within the log and the wording as follows?**

**From:**
- APPLE_PAY authorisation success

**To:**
- APPLE_PAY charge status=AUTHORISATION REJECTED (request status=success)

**QUESTION 2**
**Are we happy to fix just the log (as per Jira PP-12374) ?**
We could add a new ticket to investigate whether metricRegistry and walletPaymentAuthorisationSuccessCounter need changing too.

## How to test

- Manually, locally
- Searching in Splunk after deployment

### ~~Logging~~

- [ ] ~~only emit log lines at ERROR level which require immediate attention from a support engineer. These will trigger a zendesk alert.~~

### ~~Documentation~~

- [ ] ~~Updated README.md for any of the following ?~~

* ~~Introduced any new environment variables / removed existing environment variable~~
* ~~Added new API / updated existing API definition~~

### CODE REVIEW

**Comment by marcotranchino** [comment]
Successful payments are logged as follows:

```
[[INFO] Authorising charge for APPLE_PAY, [INFO] APPLE_PAY charge status=AUTHORISATION SUCCESS (request status=success) - charge_external_id=c92pt9r7ttfcpb56l3ha05edm4, payment provider response=Mock for WorldpayOrderStatusResponse, hashCode: 589166341, [INFO] Processing gateway auth response for APPLE_PAY, [INFO] Authorisation for c92pt9r7ttfcpb56l3ha05edm4 (sandbox transaction-id) for null (1) - Mock for WorldpayOrderStatusResponse, hashCode: 589166341 .'. AUTHORISATION SUCCESS -> AUTHORISATION SUCCESS]

```

**Comment by marcotranchino** [comment]
## Discussed with @jfharden re QUESTION 2 

(Are we happy to fix just the log (as per Jira PP-12374) ? We could add a new ticket to investigate whether metricRegistry and walletPaymentAuthorisationSuccessCounter need changing too.)

**We agreed to make the further changes in a follow up PR so that we can coordinate updating some alerts and dashboards too.**

Thanks, @jfharden !

File: src/main/java/uk/gov/pay/connector/wallets/WalletAuthoriseService.java [hunk 1]
```diff
@@ -108,8 +109,9 @@ public GatewayResponse<BaseAuthoriseResponse> authorise(String chargeId, WalletA
             Optional<Auth3dsRequiredEntity> auth3dsDetailsEntity =
                     operationResponse.getBaseResponse().flatMap(BaseAuthoriseResponse::extractAuth3dsRequiredDetails);
             CardExpiryDate cardExpiryDate = operationResponse.getBaseResponse().flatMap(BaseAuthoriseResponse::getCardExpiryDate).orElse(null);
-
-            logMetrics(charge, operationResponse, requestStatus, walletAuthorisationRequest.getWalletType());
+            
+            if (chargeStatus != null) chargeStatusValue = chargeStatus.getValue();
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/connector/wallets/WalletAuthoriseService.java]: Style could change
```suggestion
            if (chargeStatus != null) {
             chargeStatusValue = chargeStatus.getValue();
            }
```

File: src/main/java/uk/gov/pay/connector/wallets/WalletAuthoriseService.java [hunk 2]
```diff
@@ -136,10 +138,11 @@ public GatewayResponse<BaseAuthoriseResponse> authorise(String chargeId, WalletA
     private void logMetrics(ChargeEntity chargeEntity,
                             GatewayResponse<BaseAuthoriseResponse> operationResponse,
                             String successOrFailure,
+                            String chargeStatusValue,
                             WalletType walletType) {
 
-        LOGGER.info("{} authorisation {} - charge_external_id={}, payment provider response={}",
-                walletType.toString(), successOrFailure, chargeEntity.getExternalId(), operationResponse.toString());
+        LOGGER.info("{} charge status={} (request status={}) - charge_external_id={}, payment provider response={}",
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/connector/wallets/WalletAuthoriseService.java]: We should still have `authorisation` in the string. Otherwise, it won't be clear what activity is being logged for.

Comment (kbottla) [inline src/main/java/uk/gov/pay/connector/wallets/WalletAuthoriseService.java]: format of the logging can be improved by using key-value pairs. Look for usages of `kv(`.

If not using `kv`, it's probably better to use comma separate values. `(request status={})` is not the pattern we have for logs.

File: src/main/java/uk/gov/pay/connector/wallets/WalletAuthoriseService.java [hunk 3]
```diff
@@ -68,6 +68,7 @@ public GatewayResponse<BaseAuthoriseResponse> authorise(String chargeId, WalletA
             final ChargeEntity charge = prepareChargeForAuthorisation(chargeId);
             GatewayResponse<BaseAuthoriseResponse> operationResponse;
             ChargeStatus chargeStatus = null;
+            String chargeStatusValue = "N/A";
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/connector/wallets/WalletAuthoriseService.java]: Not convinced we should have `N/A` as charge status. We always get a charge status even when requests fail. Are there any instances when it is null?

Comment (marcotranchino) [reply src/main/java/uk/gov/pay/connector/wallets/WalletAuthoriseService.java]: Thanks, @kbottla !

We would have it null if the request fails to connect to the Payment Gateway, so for request status failure.

Comment (marcotranchino) [reply src/main/java/uk/gov/pay/connector/wallets/WalletAuthoriseService.java]: I could potentially avoid using a variable and pass `chargeStatus.getValue()` directly, perhaps with ternary operator:

```java
logMetrics(charge, operationResponse, requestStatus, chargeStatus != null ? chargeStatus.getValue() : null, walletAuthorisationRequest.getWalletType());
```

And then no need to initialise a variable at all. Better this way?

File: src/main/java/uk/gov/pay/connector/wallets/WalletAuthoriseService.java (hunk 1 above)
Comment (marcotranchino) [reply src/main/java/uk/gov/pay/connector/wallets/WalletAuthoriseService.java]: Thanks! I'll update this as suggested, unless we remove the variable chargeStatusValue altogether.

File: src/main/java/uk/gov/pay/connector/wallets/WalletAuthoriseService.java (hunk 2 above)
Comment (marcotranchino) [reply src/main/java/uk/gov/pay/connector/wallets/WalletAuthoriseService.java]: Adding it back now.

Comment (marcotranchino) [reply src/main/java/uk/gov/pay/connector/wallets/WalletAuthoriseService.java]: Amended. 

Example below.

```
APPLE_PAY authorisation - charge status=AUTHORISATION SUCCESS - request status=success - charge_external_id=%s, payment provider response=%s"
```

File: src/main/java/uk/gov/pay/connector/wallets/WalletAuthoriseService.java [hunk 4]
```diff
@@ -136,10 +140,11 @@ public GatewayResponse<BaseAuthoriseResponse> authorise(String chargeId, WalletA
     private void logMetrics(ChargeEntity chargeEntity,
                             GatewayResponse<BaseAuthoriseResponse> operationResponse,
                             String successOrFailure,
+                            String chargeStatusValue,
                             WalletType walletType) {
 
-        LOGGER.info("{} authorisation {} - charge_external_id={}, payment provider response={}",
-                walletType.toString(), successOrFailure, chargeEntity.getExternalId(), operationResponse.toString());
+        LOGGER.info("{} authorisation - charge status={} - request status={} - charge_external_id={}, payment provider response={}",
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/connector/wallets/WalletAuthoriseService.java]:
```suggestion
        LOGGER.info("{} authorisation - charge status={}, request status={}, charge_external_id={}, payment provider response={}",
```

File: src/main/java/uk/gov/pay/connector/wallets/WalletAuthoriseService.java (hunk 3 above)
Comment (kbottla) [reply src/main/java/uk/gov/pay/connector/wallets/WalletAuthoriseService.java]: yeah, we don't need to add a new variable, and also don't need to call `getValue()`. The toString() method on ChargeStatus returns `getValue()`. So we can pass in `chargeStatus` and use the exact name in logging.

Comment (kbottla) [reply src/main/java/uk/gov/pay/connector/wallets/WalletAuthoriseService.java]: If a request fails, then the charge status will be AUTHORISATION ERROR due to https://github.com/alphagov/pay-connector/blob/f2941db904b67b8e8cf9d0a07bb0b2be4dcc90e1/src/main/java/uk/gov/pay/connector/paymentprocessor/service/AuthorisationService.java#L100. We don't usually leave charge status as it is post an operation.

Would be good to add a test, if null is a possible value and to assert that we are handling it correctly.

Comment (marcotranchino) [reply src/main/java/uk/gov/pay/connector/wallets/WalletAuthoriseService.java]: We looked at the code together and found that chargeStatus cannot be null at this point in the code.

Comment (marcotranchino) [reply src/main/java/uk/gov/pay/connector/wallets/WalletAuthoriseService.java]: I have applied the other suggestion. Thanks again! 🙌

**Review by kbottla — APPROVED** [review]
